File: backend-dotnet/Application/Commands/Auth/RegisterUserCommandHandler.cs
using Application.Interfaces;
using BCrypt.Net;
using Domain.Entities;
using Microsoft.Extensions.Logging;

namespace Application.Commands.Auth
{
    public class RegisterUserCommandHandler : ICommandHandler<RegisterUserCommand, (string token, User user)>
    {
        private readonly IUserRepository _userRepository;
        private readonly IJwtService _jwtService;
        private readonly ICacheService _cacheService;
        private readonly ILogger<RegisterUserCommandHandler> _logger;

        public RegisterUserCommandHandler(
            IUserRepository userRepository,
            IJwtService jwtService,
            ICacheService cacheService,
            ILogger<RegisterUserCommandHandler> logger)
        {
            _userRepository = userRepository;
            _jwtService = jwtService;
            _cacheService = cacheService;
            _logger = logger;
        }

        public async Task<(string token, User user)> HandleAsync(RegisterUserCommand command)
        {
            var dto = command.Dto;
            
            _logger.LogInformation("Registering user with email: {Email}", dto.Email);

            if (await _userRepository.ExistsByEmailAsync(dto.Email))
            {
                _logger.LogWarning("Registration failed: Email {Email} already exists", dto.Email);
                throw new ArgumentException("This email is already in use.");
            }

            if (await _userRepository.ExistsByUsernameAsync(dto.Username))
            {
                _logger.LogWarning("Registration failed: Username {Username} already exists", dto.Username);
                throw new ArgumentException("This username is already taken.");
            }

            var user = new User
            {
                Email = dto.Email,
                Username = dto.Username,
                PasswordHash = BCrypt.Net.BCrypt.HashPassword(dto.Password)
            };

            _userRepository.Add(user);
            await _userRepository.SaveChangesAsync();

            await InvalidateUserAutocompleteCaches(user.Username);

            var token = _jwtService.GenerateToken(user);
            
            _logger.LogInformation("User {UserId} registered successfully with username: {Username}", 
                user.Id, user.Username);

            return (token, user);
        }

        private Task InvalidateUserAutocompleteCaches(string username)
        {
            try
            {
                var normalizedUsername = username.ToLower();

                if (normalizedUsername.Length >= 2)
                {
                    var firstTwoChars = normalizedUsername.Substring(0, 2);
                    _cacheService.RemoveByPattern($"autocomplete:{firstTwoChars}*");
                }

                if (normalizedUsername.Length >= 1)
                {
                    var firstChar = normalizedUsername.Substring(0, 1);
                    _cacheService.RemoveByPattern($"autocomplete:{firstChar}*");
                }

                _logger.LogDebug("Invalidated autocomplete caches for new user: {Username}", username);
            }
            catch (Exception ex)
            {
                _logger.LogWarning(ex, "Failed to invalidate autocomplete caches for user: {Username}", username);
            }

            return Task.CompletedTask;
        }
    }
}

File: backend-java/src/main/java/com/jerrygram/application/commands/auth/RegisterUserCommandHandler.java
package com.jerrygram.application.commands.auth;

import com.jerrygram.application.common.AuthResult;
import com.jerrygram.application.common.UserIndex;
import com.jerrygram.application.interfaces.ICacheService;
import com.jerrygram.application.interfaces.ICommandHandler;
import com.jerrygram.application.interfaces.IElasticService;
import com.jerrygram.application.interfaces.IJwtService;
import com.jerrygram.domain.entities.User;
import com.jerrygram.infrastructure.repositories.UserRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
@Slf4j
public class RegisterUserCommandHandler implements ICommandHandler<RegisterUserCommand, AuthResult> {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final IJwtService jwtService;
    private final ICacheService cacheService;
    private final IElasticService elasticService;

    @Override
    @Transactional
    public AuthResult handle(RegisterUserCommand command) {
        try {
            log.info("=== STEP 1: Starting registration process ===");
            var dto = command.getRegisterDto();
            
            log.info("=== STEP 2: Got DTO, email: {} ===", dto.getEmail());

            // Create new user using builder to avoid initialization issues
            log.info("=== STEP 3: Creating new user with builder ===");
            var user = User.builder()
                    .email(dto.getEmail())
                    .username(dto.getUsername())
                    .passwordHash(passwordEncoder.encode(dto.getPassword()))
                    .build();

            log.info("=== STEP 4: Saving user to database ===");
            userRepository.save(user);
            log.info("=== STEP 4 SUCCESS: User saved with ID: {} ===", user.getId());

            // Generate JWT token
            log.info("=== STEP 5: Generating JWT token ===");
            var token = jwtService.generateToken(user);
            log.info("=== STEP 5 SUCCESS: Token generated ===");
            
            log.info("=== STEP 6: Building AuthResult ===");
            var userInfo = AuthResult.UserInfo.builder()
                    .id(user.getId())
                    .username(user.getUsername())
                    .email(user.getEmail())
                    .profileImageUrl(user.getProfileImageUrl())
                    .createdAt(user.getCreatedAt())
                    .build();
                    
            var result = AuthResult.builder()
                    .token(token)
                    .user(userInfo)
                    .build();
            
            log.info("=== SUCCESS: User {} registered successfully ===", user.getUsername());
            return result;
        } catch (Exception e) {
            log.error("=== ERROR in registration process ===", e);
            throw e;
        }
    }

    private void invalidateUserAutocompleteCaches(String username) {
        try {
            var normalizedUsername = username.toLowerCase();

            if (normalizedUsername.length() >= 2) {
                var firstTwoChars = normalizedUsername.substring(0, 2);
                cacheService.deleteByPattern("autocomplete:" + firstTwoChars + "*");
            }

            if (normalizedUsername.length() >= 1) {
                var firstChar = normalizedUsername.substring(0, 1);
                cacheService.deleteByPattern("autocomplete:" + firstChar + "*");
            }

            log.debug("Invalidated autocomplete caches for new user: {}", username);
        } catch (Exception ex) {
            log.warn("Failed to invalidate autocomplete caches for user: {}", username, ex);
        }
    }
}